syntax = "proto3";

package erebus;


message PropertyInfo {
  uint32 id = 1;
  uint32 type = 2;
  string name = 3;
  string readableName = 4;  
}

message Property {
  uint32 id = 1;
  oneof value {
    bool v_bool = 2;
    int32 v_int32 = 3;
    uint32 v_uint32 = 4;
    int64 v_int64 = 5;
    uint64 v_uint64 = 6;
    double v_double = 7;
    string v_string = 8;
    bytes v_binary = 9;
  }
}

message Exception {
  optional string message = 1;
  repeated Property props = 2;
}

service Erebus {
  rpc GenericCall(ServiceRequest) returns(ServiceReply) {}
  rpc GenericStream(ServiceRequest) returns(stream ServiceReply) {}
}

message ServiceRequest {
  string request = 1;
  uint32 clientId = 2;
  repeated Property args = 3;
  repeated PropertyInfo mapping = 4;
  optional uint64 timestamp = 5;
}

message ServiceReply {
  optional Exception exception = 1;  
  repeated Property props = 2;
  repeated PropertyInfo mapping = 3;
  optional uint64 timestamp = 4;         // unchanged timestamp from ServiceRequest
  optional uint64 duration = 5;          // real request processing time
}
